File: modularity_conn/84.composite_set_preserved_group_iters.R
## read set preserved and get median at every voxel. 
## goes across the group data, so uses tlrc space data
## thus doing this in batches

## !!!!!!! THIS IS MUCH LIKE '61.set_members_avg_preserved.R'. 
## Different bcs getting median image for the 'iters' which are the null (condition against itself) and minor coding diffs
## Above line is no longer true. Now using this to get for real participant data, i.e., true 'SNSC' across group

fltmedian <- function(x) if (777 %in% x) return(777) else return(median(x))   # function to keep row with value of 777 if it shows up in any row.
fltmean <- function(x) if (777 %in% x) return(777) else return(mean(x))

subjects <- c("ANGO","CLFR","MYTP","TRCO","PIGL","SNNW","LDMW","FLTM","EEPA","DNLN","CRFO","ANMS","MRZM","MRVV","MRMK","MRMC","MRAG","MNGO","LRVN")
nvoxels <- 231203
null_mat <- matrix(nrow = nvoxels, ncol = length(subjects))

cc <- c()
for (ss in 1:length(subjects))
{
    #ss_null <- paste("/mnt/tier2/urihas/Andric/steadystate/",subjects[ss],"/corrTRIM_BLUR/preserved_",subjects[ss],"_median5p_20vxFltr_warped_tlrc_dump.txt", sep = "")
    ss_null <- paste("/mnt/tier2/urihas/Andric/steadystate/",subjects[ss],"/corrTRIM_BLUR/iters_",subjects[ss],"_median5p_20vxFltr_warped_tlrc_dump.txt", sep = "")
    null_mat[,ss] <- scan(ss_null, quiet = TRUE)[seq(4, nvoxels*4, 4)]
    cc <- c(cc, length(which(null_mat[,ss] == 777))/nvoxels)
}

print(paste("Percent of vovxels in each column (each subject) that are 777"))
print(cc * 100)
median_vec <- apply(null_mat, 1, fltmedian)   # applying function to mark voxel as 777 if any voxel across subjects is 777. 777 carries from voxels in modules where n < 20
average_vec <- apply(null_mat, 1, fltmean)

print(paste(round(((length(which(median_vec == 777)) / nvoxels) * 100), 4)," % of voxels in median group vec that are 777 ", sep = ""))
print(paste(round(((length(which(average_vec == 777)) / nvoxels) * 100), 4)," % of voxels in average group vec that are 777 ", sep = ""))


setwd("/mnt/tier2/urihas/Andric/steadystate/groupstats")

write.table(round(median_vec, 4), "iters_group_median5p_20vxFltr_warped_median.txt", row.names = F, col.names = F, quote = F)
write.table(round(average_vec, 4), "iters_group_median5p_20vxFltr_warped_average.txt", row.names = F, col.names = F, quote = F)

File: modularity_conn/84.composite_set_preserved_group_iters.py
#!/usr/bin/python
"""
This is a python port of R script by the same name
"""
import sys
import os
import numpy as np
from time import ctime

def fltmedian(nr):
    if len(nr[nr==777.]) / float(len(nr)) >= .5:
        return 777
    else:
        return np.median(nr[nr!=777])

def fltmean(nr):
    if len(nr[nr==777.]) / float(len(nr)) >= .5:
        return 777
    else:
        return np.mean(nr[nr!=777]) 



if __name__ == "__main__":

    subjects = ["ANGO", "MYTP", "TRCO", "PIGL", "SNNW", "LDMW", "FLTM", "CLFR", "EEPA", "DNLN", "CRFO", "ANMS", "MRZM", "MRVV", "MRMK", "MRMC", "MRAG", "MNGO", "LRVN"]
    nvoxels = 231203
    null_mat = np.array(np.zeros(nvoxels*len(subjects))).reshape(nvoxels, len(subjects))

    for i in xrange(len(subjects)):
        print subjects[i]+" // "+ctime()
        #ss_nullfname = os.environ['state']+'/%s/corrTRIM_BLUR/iters_%s_median5p_20vxFltr_warped_tlrc_dump.txt' % (subjects[i], subjects[i])
        ss_nullfname = os.environ['state']+'/%s/corrTRIM_BLUR/preserved_%s_median5p_20vxFltr_warped_tlrc_dump.txt' % (subjects[i], subjects[i])
        ss_nullf = open(''.join(ss_nullfname))
        ss_null = (line.split()[3] for line in ss_nullf)
        null_mat[:,i] = map(float, ss_null)

    outmedian = np.apply_along_axis(fltmedian, axis = 1, arr = null_mat)
    outmean = np.apply_along_axis(fltmean, axis = 1, arr = null_mat)

    os.chdir(os.environ['state']+'/groupstats')
    #np.savetxt('iters_group_median5p_20vxFltr_warped_median.out', outmedian, fmt='%.4f')
    #np.savetxt('iters_group_median5p_20vxFltr_warped_average.out', outmean, fmt='%.4f')
    np.savetxt('preserved_group_median5p_20vxFltr_warped_median.out', outmedian, fmt='%.4f')
    np.savetxt('preserved_group_median5p_20vxFltr_warped_average.out', outmean, fmt='%.4f')

    m1 = "Voxels in median group that are 777: %s" % round((len(outmedian[outmedian==777]) / nvoxels) * 100, 4)
    m2 = "Voxels in mean group that are 777: %s" % round((len(outmean[outmean==777]) / nvoxels) * 100, 4)
    fsummary = open('preserved_group_median5p_20vxFltr_warped_summary.out', 'w')
    fsummary.write(m1+'\n'+m2+'\n')
    fsummary.close()

